### PR TITLE
Feature/file encryption

### PR DESCRIPTION
**Pull request by tomShoutTheSecond**
Encryption implementation using X25519 key pair generator and Threefish cipher
- pair devices for encrypted communication
- encrypt files before sending
- decrypt files after receiving

See EncryptionService, particularly the documentation at the top, for implementation details

### CODE REVIEW

**Comment by wellhat** [comment]
Really awesome work @tomShoutTheSecond, love what you made out of my feedback on the pairing screen, absolutely 🔥

This PR does break image previews, multiple file sharing, and a couple more features (understandably so). It would be good to try and improve the UX so that we're not back to downloading .zips. Atm the only option which works for decrypting encrypted files is "Save" which means going outside the SkyDrop app to find the downloaded files.

**Comment by wellhat** [comment]
The icon showing "Anyone with a link" or "Contact name" should have icons which make this clear, maybe look at what they do on Google Docs

**Comment by tomShoutTheSecond** [comment]
Updates:
x take List<Contact> for recipients
x change checkmark icon iOS
x impose a limit on max recipients count
x put filename in encrypted file?
x add new icon for .skydrop files
x add icon to recipient button
x allow sender to decrypt their own files
x add name to device, so user doesn't have to input contact name when pairing
    x share device name in pairing QR code
    x sanitise names to ascii
    x limit name length
    x enable changing device name from settings
x upload encrypted files as extensionless guid with hyphens stripped
x place a signature in the middle of the guid filename to identify encrypted single file, or encrypted zip
x ensure that contact name collisions are handled correctly, by adding (1), (2), (3) etc. to the contact name
x implement error overlay for pairing screen (easy to test it with invalid QR)
x add rename option for contacts
x show unzip button for encrypted zip files, enable unzipping in app
x add 2 byte version info to pairing QR codes
x add "no contacts" label to Android
x randomise archive names

**Comment by wellhat** [comment]
@tomShoutTheSecond besides my comment about the security with Realm storage, I think this is ready to merge (although I haven't tested most recent changes yet). There are some minor refactors as well but that's the primary thing I wanted to raise.

**Comment by wellhat** [comment]
There's also conflicts @tomShoutTheSecond

File: src/SkyDrop.Core/Services/StorageService.cs [hunk 1]
```diff
@@ -161,6 +162,76 @@ public void SetAverageUploadRate(UploadAverage uploadAverage)
...
+            });
+        }
+
+        public (bool exists, string savedName) ContactExists(Guid id)
+        {
+            var existingContact = realm.Find<ContactRealmObject>(id.ToString());
+            var exists = existingContact != null;
+            return (exists, existingContact?.Name);
+        }
+
```
[the remark this answers is not in the document]
Comment (tomShoutTheSecond) [reply src/SkyDrop.Core/Services/StorageService.cs:213]: Yes you can rename from contacts screen

**Pull request merged**